Add post about libcurl4/libxml2 dependency errors (#60)
vapor/blog: Content/posts/error-while-loading-shared-libs.md

---
date: 2099-07-18 09:00
description: As little as possible, but no less
tags: TBD

> author: Galen

Reviewer: Can you add a new field below called `author_username` that has your GitHub username and one for `author_twitter` for Twitter if you want to have those linked. We don't use it yet, but we will soon

Author: I'm not active on Twitter, so I'm happy to leave that to the next guest author.

How about `author_github` instead of `author_username`?

Reviewer: Yeah that makes more sense

---

> # Docker image: Error while loading shared libraries: libcurl4 or libxml2

Author: How about title/description:

> ## Dockerfile makes adding optional Swift dependencies easier
> A simple fix for when you happen upon an "error while loading shared libraries" libcurl4 or libxml2
<img width="859" alt="CleanShot 2022-07-19 at 11 33 26@2x" src="https://user-images.githubusercontent.com/3375/179823861-c8f81a52-4345-4fa5-9c41-d7c6de70fa77.png">

<img width="871" alt="CleanShot 2022-07-19 at 11 33 44@2x" src="https://user-images.githubusercontent.com/3375/179823873-2f05b2e4-aceb-4f32-a158-e9b23a3246c1.png">

> # Docker image: Error while loading shared libraries: libcurl4 or libxml2

Reviewer: Looks good

FROM ubuntu:focal

# Make sure all system packages are up to date, and install only essential packages.
RUN export DEBIAN_FRONTEND=noninteractive DEBCONF_NONINTERACTIVE_SEEN=true \
    && apt-get -q update \
    && apt-get -q dist-upgrade -y \
    && apt-get -q install -y \
      ca-certificates \
      tzdata \
# If your app or its dependencies import FoundationNetworking, also install `libcurl4`.
      # libcurl4 \
# If your app or its dependencies import FoundationXML, also install `libxml2`.
      # libxml2 \
    && rm -r /var/lib/apt/lists/*
[…]
```

When your app executable has been linked to the dynamic library for either `libcurl4` or `libxml2` but they aren't installed, your app will crash immediately on load with one of the following errors:

> ./Run: error while loading shared libraries: libcurl.so.4: cannot open shared object file: No such file or directory

> ./Run: error while loading shared libraries: libxml2.so.2: cannot open shared object file: No such file or directory

* If your app or its dependencies import `FoundationNetworking`, you'll need `libcurl4`.
* If your app or its dependencies import `FoundationXML`, you'll need `libxml2`.

> Start by omitting both `libcurl4` and `libxml2` if you're unsure that your app needs them. Then, build and run your app image. Your app will crash immediately on load if either or both are needed. (You can also inspect your binary's dynamic dependencies using `ldd`.)

Reviewer: Refine to remove repetition
```suggestion
Start by omitting both `libcurl4` and `libxml2` if you're unsure that your app needs them. Then, build and run your app image. If it crashes you'll need to add either or both. (You can also inspect your binary's dynamic dependencies using `ldd`.)
```
